### PR TITLE
Add #[repr(transparent)] to some libcore types

### PR DESCRIPTION
**Pull request by SimonSapin**
* `UnsafeCell`
* `Cell`
* `NonZero*`
* `NonNull`
* `Unique`

CC https://github.com/rust-lang/rust/issues/43036

### CODE REVIEW

**Comment by SimonSapin** [comment]
Let’s get libs and lang team sign off for libcore types:

@rfcbot fcp merge

Are there more types that belong on this list?

**Comment by willmo** [comment]
Adding it to the atomic types would be very helpful, since it would allow them to be used in `#[repr(C)]` structs.

**Comment by eddyb** [comment]
Do we guarantee the size of atomic sizes, btw?

**Comment by SimonSapin** [comment]
@willmo Good point, added.

@eddyb The current implementation has the size of each atomic type be what the name suggests, and makes some of them not available depending on the [`Target::max_atomic_width`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/spec/struct.Target.html#method.max_atomic_width)

**Comment by fitzgen** [comment]
IIRC, C++ will use mutexes and/or wider integers to simulate support when the target doesn't support the atomic's width. Are we sure that we don't intend to ever do that as well? Because giving guarantees about the size of an atomic would make that impossible, unless I am missing something.

File: src/libcore/sync/atomic.rs [hunk 1]
```diff
@@ -124,6 +124,7 @@ pub fn spin_loop_hint() {
 /// [`bool`]: ../../../std/primitive.bool.html
 #[cfg(target_has_atomic = "8")]
 #[stable(feature = "rust1", since = "1.0.0")]
+#[repr(transparent)]
 pub struct AtomicBool {
```
Comment (sfackler) [inline src/libcore/sync/atomic.rs]: This is a bit iffy I think. Presumably you'd expect it to be FFI compatible with a bool, not a u8, right?

Comment (eddyb) [reply src/libcore/sync/atomic.rs]: Do we not have `AtomicBool` on platforms without `u8` atomics? I could've sworn we were using `usize`  before, which would've made the `AtomicBool` change a breaking one - assuming there is a platform where `usize` atomics exist but not byte ones.

Also, why is this not wrapping `AtomicU8`? Seems like the code would be safer/cleaner that way.

**Comment by glandium** [comment]
In C++ std::atomic<bool> is not necessarily the size of a bool.

**Comment by eddyb** [comment]
@sfackler What if we did the "mutex" thing but through a one-byte spinning "rw-lock" with no OS handoff for sleeping? It would work for bare-metal usecases just as well, assuming the types it's used with are small enough to not melt one core while another operates on the data.

**Comment by glandium** [comment]
@eddyb that wouldn't be "transparent".

**Comment by joshtriplett** [comment]
I would argue that the atomic types should have the appropriate native size, and not exist if that type can't be atomic on the target platform. If we want a type based on a mutex, that should use a different structure that can embed some form of synchronization.

File: src/libcore/sync/atomic.rs (hunk 1 above)
Comment (sfackler) [reply src/libcore/sync/atomic.rs]: The type has a `#[cfg(target_has_atomic = "8")]`. It should start wrapping AtomicU8, but it existed way before that type did.

**Comment by willmo** [comment]
@SimonSapin I certainly understand why AtomicBool is contentious/confusing, but I suspect it's less common/useful for FFI, and I certainly don't need it for FFI. So I don't mind losing that one.

But it would be a shame to lose out on `#[repr(transparent)]` for the atomic types that use the `atomic_int` macro. ISTM that their sizes *are* currently well-defined, because currently they are only available on platforms that natively support them (like @joshtriplett suggested), thanks to `#[cfg(target_has_atomic)]`.

If/when Rust needs to support atomic ints of sizes that aren't natively supported on a given platform, it seems like they could either use an entirely different type like @joshtriplett suggested, or be made not `#[repr(transparent)]` and therefore not usable in `#[repr(C)]` structs on those platforms. (I don't know if conditional `repr` is possible or desirable in current Rust.)

Otherwise, if no atomic type can ever be `#[repr(transparent)]`, I'd be curious how people want Rust to support atomically manipulating memory shared with C without writing assembly or calling into C.

**Comment by SimonSapin** [comment]
@willmo It’s only removed *from this PR*. Feel free to open another PR that can be debated without blocking the stabilization of the attribute itself.

**Comment by SimonSapin** [comment]
I’ve landed the attribute stabilization separately at https://github.com/rust-lang/rust/pull/51562, this is now only about adding it to some types.

**Comment by SimonSapin** [comment]
> `Failed to open TCP connection to rust-lang-ci2.s3.us-west-1.amazonaws.com:443 (Network is unreachable - connect(2) for "rust-lang-ci2.s3.us-west-1.amazonaws.com" port 443) (Errno::ENETUNREACH)`

Flaky network?

@bors retry

**Comment by willmo** [comment]
Re: `#[repr(transparent)]` on atomics:

> Feel free to open another PR that can be debated without blocking the stabilization of the attribute itself.

Done: #52149 :-)

**Pull request merged**